File: src/utilities.test.ts
import { test } from "node:test";
import { expect } from "expect";

import { forEachChild, parseSourceFile } from "./parser.ts";
import { isValidIdentifier, skipTrivia } from "./utilities.ts";
import { type Identifier, type Node, SyntaxKind } from "./types.ts";

test("isValidIdentifier accepts names and rejects keywords/garbage", () => {
  expect(isValidIdentifier("foo")).toBe(true);
  expect(isValidIdentifier("_x$2")).toBe(true);
  expect(isValidIdentifier("$")).toBe(true);
  expect(isValidIdentifier("2foo")).toBe(false); // leading digit
  expect(isValidIdentifier("a b")).toBe(false); // space
  expect(isValidIdentifier("")).toBe(false);
  expect(isValidIdentifier("class")).toBe(false); // reserved keyword
  expect(isValidIdentifier("true")).toBe(false); // reserved literal
});

test("skipTrivia advances over whitespace and line breaks", () => {
  expect(skipTrivia("   x", 0)).toBe(3);
  expect(skipTrivia("\n\t x", 0)).toBe(3);
  expect(skipTrivia("x", 0)).toBe(0);
});

test("skipTrivia advances over line and block comments", () => {
  expect(skipTrivia("// note\nx", 0)).toBe(8);
  expect(skipTrivia("/* note */ x", 0)).toBe(11);
  expect(skipTrivia("  /*a*/ /*b*/  x", 0)).toBe(15);
});

// The real fix: an identifier node's pos includes leading trivia (the full
// start, as in the TS compiler). skipTrivia maps it to the token's real start,
// which is the range a goto/reference/rename highlight should use.
test("an identifier node's range is trimmed to the name by skipTrivia", () => {
  const text = "class C {\n    int field;\n    int m() { return field; }\n}";
  const sf = parseSourceFile("T.java", text);

  // find the `field` reference inside the return statement (the second occurrence)
  const ids: Identifier[] = [];
  const walk = (n: Node): void => {
    if (n.kind === SyntaxKind.Identifier && (n as Identifier).text === "field") {
      ids.push(n as Identifier);
    }
    forEachChild(n, c => {
      walk(c);
      return undefined;
    });
  };
  walk(sf);
  const useNode = ids[ids.length - 1]!;

  // node.pos sits on whitespace before the name; skipTrivia lands on the name.
  expect(text[useNode.pos]).toMatch(/\s/);
  const start = skipTrivia(text, useNode.pos);
  expect(text.slice(start, useNode.end)).toBe("field");
});

File: togo/internal/compiler/utilities_test.go
package compiler

import (
	"regexp"
	"testing"
)

// Port of src/compiler/utilities.test.ts.

func TestIsValidIdentifier(t *testing.T) {
	cases := map[string]bool{
		"foo": true, "_x$2": true, "$": true,
		"2foo": false, "a b": false, "": false, "class": false, "true": false,
	}
	for name, want := range cases {
		if got := IsValidIdentifier(name); got != want {
			t.Errorf("IsValidIdentifier(%q) = %v, want %v", name, got, want)
		}
	}
}

func TestSkipTriviaWhitespace(t *testing.T) {
	cases := []struct {
		text string
		want int
	}{
		{"   x", 3}, {"\n\t x", 3}, {"x", 0},
	}
	for _, tc := range cases {
		if got := SkipTrivia(tc.text, 0); got != tc.want {
			t.Errorf("SkipTrivia(%q) = %d, want %d", tc.text, got, tc.want)
		}
	}
}

func TestSkipTriviaComments(t *testing.T) {
	cases := []struct {
		text string
		want int
	}{
		{"// note\nx", 8}, {"/* note */ x", 11}, {"  /*a*/ /*b*/  x", 15},
	}
	for _, tc := range cases {
		if got := SkipTrivia(tc.text, 0); got != tc.want {
			t.Errorf("SkipTrivia(%q) = %d, want %d", tc.text, got, tc.want)
		}
	}
}

func TestIdentifierRangeTrimmedBySkipTrivia(t *testing.T) {
	text := "class C {\n    int field;\n    int m() { return field; }\n}"
	sf := ParseSourceFile("T.java", text)
	var ids []*Node
	var walk Visitor
	walk = func(n *Node) bool {
		if n.Kind == Identifier && n.AsIdentifier().Text == "field" {
			ids = append(ids, n)
		}
		n.ForEachChild(walk)
		return false
	}
	sf.ForEachChild(walk)
	useNode := ids[len(ids)-1]
	if !regexp.MustCompile(`\s`).MatchString(string(text[useNode.Pos])) {
		t.Error("node.pos should sit on whitespace before the name")
	}
	start := SkipTrivia(text, useNode.Pos)
	if got := text[start:useNode.End]; got != "field" {
		t.Errorf("trimmed range = %q, want field", got)
	}
}
